src/Middleware/HeaderPropagation/src/DependencyInjection/HeaderPropagationHttpClientBuilderExtensions.cs

```diff
@@ -22,7 +23,48 @@ public static IHttpClientBuilder AddHeaderPropagation(this IHttpClientBuilder bu
 
             builder.Services.AddHeaderPropagation();
 
-            builder.AddHttpMessageHandler<HeaderPropagationMessageHandler>();
+            builder.AddHttpMessageHandler(services =>
+            {
+                var options = new HeaderPropagationClientOptions();
+                var middlewareOptions = services.GetRequiredService<IOptions<HeaderPropagationOptions>>();
+                for (int i = 0; i < middlewareOptions.Value.Headers.Count; i++)
+                {
+                    var header = middlewareOptions.Value.Headers[i];
+                    options.Headers.Add(header.CapturedHeaderName, header.CapturedHeaderName);
+                }
+                return new HeaderPropagationMessageHandler(Options.Options.Create(options), services.GetRequiredService<HeaderPropagationValues>());
+            });
+
+            return builder;
+        }
+
+        /// <summary>
+        /// Adds a message handler for propagating headers collected by the <see cref="HeaderPropagationMiddleware"/> to a outgoing request,
+        /// explicitly specifying which headers to propagate.
+        /// </summary>
+        /// <param name="builder">The <see cref="IHttpClientBuilder"/> to add the message handler to.</param>
+        /// <param name="configureOptions">A delegate used to configure the <see cref="HeaderPropagationOptions"/>.</param>
+        /// <returns>The <see cref="IHttpClientBuilder"/> so that additional calls can be chained.</returns>
+        public static IHttpClientBuilder AddHeaderPropagation(this IHttpClientBuilder builder, Action<HeaderPropagationClientOptions> configureOptions)
+        {
+            if (builder == null)
+            {
+                throw new ArgumentNullException(nameof(builder));
+            }
+
+            if (configureOptions == null)
+            {
+                throw new ArgumentNullException(nameof(configureOptions));
+            }
+
+            builder.Services.AddHeaderPropagation();
+
+            builder.AddHttpMessageHandler(services=>
+            {
+                var options = new HeaderPropagationClientOptions();
+                configureOptions(options);
+                return new HeaderPropagationMessageHandler(Options.Options.Create(options), services.GetRequiredService<HeaderPropagationValues>());
```

Comment:
Based on the usage here, there probably isn't much of a reason to use `IOptions<>` at all for `HeaderPropagationClientOptions` - in both cases we're not getting the options from DI. We could try to change it so that it does play with `IOptions<>` in DI, but it's going to be messy and I'm not sure what it helps.

Reply:
Agree. It would need something like AddHttpMessageHandler<> to automatically inject the builder name if the constructor accept a string to doing it cleanly.